drivers/adc/adc_stm32.c

```diff
@@ -277,6 +301,83 @@ struct adc_stm32_cfg {
 static bool init_irq = true;
 #endif
 
+#ifdef CONFIG_ADC_STM32_DMA
+static int adc_stm32_dma_start(const struct device *dev,
+			       void *buffer, size_t channel_count)
+{
+	const struct adc_stm32_cfg *config = dev->config;
+	ADC_TypeDef *adc = (ADC_TypeDef *)config->base;
+	struct adc_stm32_data *data = dev->data;
+	struct dma_block_config *blk_cfg;
+	int ret;
+
+	struct stream *dma = &data->dma;
+
+	blk_cfg = &dma->dma_blk_cfg;
+
+	/* Write cached memory back into SRAM. In the DMA callback
+	 * the cache is invalidated and which could possibly allow
+	 * recently written data to be lost.
+	 */
+	SCB_CleanDCache_by_Addr((uint32_t *)buffer, data->channel_count * sizeof(int16_t));
```

Comment:
Hopefully can be removed if using RAM_NOCACHE on SRAM2

Reply:
I'm not sure if this is the right approach. It would mean I should probably add a check in the dma driver to verify that the given source and destination addresses are both in nocache regions, and fail if it's not. That might break existing setups that is using cache maintenance operations.

The reason I think it's okay for the BDMA in #53184 is because the SRAM4 it's a dedicated smallish section, that's typically not used. However, an user might see `NOCACHE` is required, then add the flag to the main SRAM section used by the software, and then loose the performance benifets of having cache.

Also, the description for [CONFIG_NOCACHE_MEMORY](https://docs.zephyrproject.org/2.7.0/reference/kconfig/CONFIG_NOCACHE_MEMORY.html) seems to lean towards attempting to use regular cache maintenance operations before adding the NOCACHE flag to an entire SRAM region.
```
This memory section can be used to perform DMA
transfers when cache coherence issues are not optimal or can not
be solved using cache maintenance operations
```

It would still be useful and workable though. In our application I would put both our ADC1 and ADC3 buffers inside SRAM4, which isn't used by the software anyway.

What do you think?

Reply:
Right, I forgot this change isn't tightly linked to BDMA or H7. On other series, this would not apply.

Optionally, we can think of a Kconfig property (/dts ?) which would allow user to chose one of both approaches.

Reply:
I can add a kconfig setting to `adc/Kconfig.stm32`, and let it only be available if the device has cache, and dma is enabled, etc. Something like:
```kconfig
if ADC_STM32_DMA
config ADC_STM32_DMA_ACTIVE_CACHE_HANDLING
	bool "STM32 ADC DMA Active Cache Handling"
        default n
	help
	  If enabled the ADC driver will do manual cache maintainence
          to the output buffer which means the buffer does not need to
          be located in a NOCACHE memory region. If not enabled the
          driver will not allow buffers not located in a NOCACHE memory
          region.
endif
```

It could be added to the dma kconfig, but I think the responsibility of handling the cache is of the thing using the dma, not the dma itself.

PS. Do you know of a way to check if an address is in a NOCACHE memory region during runtime? I'm not sure how to do that. I also asked on Discord in the `#DMA` channel, maybe we can chat there?

Reply:
> It could be added to the dma kconfig, but I think the responsibility of handling the cache is of the thing using the dma, not the dma itself.

Looks fine to me

> Do you know of a way to check if an address is in a NOCACHE memory region during runtime?

Nope

Reply:
It doesn't seem possible with the current Zephyr API, or easily implementable (by me at least), to determine if a given address in RAM is non-cachable. Therefore, I'm not quite sure where to go from here.

Maybe we can keep this as-is for now?

Reply:
After a conversation with erwango on Discord it was decided to remove the manual cache maintenance. Rather the driver will expect the buffer to be in a non-cachable SRAM region. This will be enfored by doing runtime check on the buffer supplied when a read is started, but only for the `STM32H7`. This check will only pass if the buffer is placed in a DTS SRAM region which is marked as `NON_CACHEABLE`, meaning the `.nocache` section will not work. It seems like using DTS section is the intended anyway.

Reply:
Updated as described in previous comment.